### PR TITLE
[Search] Update SDK with new REST spec - 7601061b

### LINKED ISSUE #20809: [search-documents]Add 3 new V3 Skills to the search-documents SDK
The search-documents swagger will be introducing 3 new skills to their service.

1. SentimentV3
2. EntityLinkingV3
3. EnttyRecognitionV3

The Swagger changes to introduce these new skills is at  https://github.com/Azure/azure-rest-api-specs-pr/pull/3395. Per the current plan these new skills will be globally available after May 17, 2021. 

The SDK Changes should be completed for the June Release

### PR DESCRIPTION
**Pull request by Mohit-Chakraborty**
REST spec -
* https://raw.githubusercontent.com/Azure/azure-rest-api-specs/7601061bfbb72b2f19cb29c46cbf9a397c2d8893/specification/search/data-plane/Azure.Search/preview/2021-04-30-Preview/searchindex.json
* https://raw.githubusercontent.com/Azure/azure-rest-api-specs/7601061bfbb72b2f19cb29c46cbf9a397c2d8893/specification/search/data-plane/Azure.Search/preview/2021-04-30-Preview/searchservice.json

### CODE REVIEW

**Comment by Mohit-Chakraborty** [comment]
cc: @tg-msft

**Review by tg-msft — COMMENTED** [review]
Quick comments on the API shape

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs [hunk 1]
```diff
@@ -680,6 +692,23 @@ public partial class EntityRecognitionSkill : Azure.Search.Documents.Indexes.Mod
         public static bool operator !=(Azure.Search.Documents.Indexes.Models.EntityRecognitionSkillLanguage left, Azure.Search.Documents.Indexes.Models.EntityRecognitionSkillLanguage right) { throw null; }
         public override string ToString() { throw null; }
     }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public readonly partial struct EntityRecognitionSkillVersion : System.IEquatable<Azure.Search.Documents.Indexes.Models.EntityRecognitionSkillVersion>
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs]: Done. I wonder if I should rename both `EntityRecognitionSkillVersion` and `SentimentSkillVersion` to `SkillVersion` or even just `Version`.
With the nesting they don't need the long names anymore.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs [hunk 2]
```diff
@@ -1512,6 +1562,35 @@ public partial class PhoneticTokenFilter : Azure.Search.Documents.Indexes.Models
...
+        public string DefaultLanguageCode { get { throw null; } set { } }
+        public string Domain { get { throw null; } set { } }
+        public string MaskingCharacter { get { throw null; } set { } }
+        public Azure.Search.Documents.Indexes.Models.PIIDetectionSkillMaskingMode? MaskingMode { get { throw null; } set { } }
+        public double? MinimumPrecision { get { throw null; } set { } }
+        public string ModelVersion { get { throw null; } set { } }
+        public System.Collections.Generic.IList<string> PiiCategories { get { throw null; } }
+    }
+    [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+    public readonly partial struct PIIDetectionSkillMaskingMode : System.IEquatable<Azure.Search.Documents.Indexes.Models.PIIDetectionSkillMaskingMode>
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs]: > Why is PII cased differently than PiiDetectionSkill?

I renamed `PIIDetectionSkill`->`PiiDetectionSkill` to conform to naming guidelines. Missed this other instance. Have changed the other one now too.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs [hunk 3]
```diff
@@ -1667,6 +1746,14 @@ public partial class SearchIndexerDataContainer
         public string Name { get { throw null; } set { } }
         public string Query { get { throw null; } set { } }
     }
+    public partial class SearchIndexerDataIdentity
+    {
+        public SearchIndexerDataIdentity() { }
+    }
+    public partial class SearchIndexerDataNoneIdentity : Azure.Search.Documents.Indexes.Models.SearchIndexerDataIdentity
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:1755]: This is from [the spec](https://raw.githubusercontent.com/Azure/azure-rest-api-specs/7601061bfbb72b2f19cb29c46cbf9a397c2d8893/specification/search/data-plane/Azure.Search/preview/2021-04-30-Preview/searchservice.json).

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs (hunk 1 above)
[the remark this answers is not in the document]
Comment (tg-msft) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs]: I like `SkillVersion` to match `ServiceVersion` but still not confuse things with `ModelVersion`.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (tg-msft) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:1755]: We should really spend some time here before we get anywhere close to GA.  I think some of these are really confusing names (i.e., I create an instance of this type to clear the value of a property?) in one of the hardest to understand areas.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs]: > Do we need Skill in this name? Could we just call it PiiDetectionMaskingMode?

It think it helps to tie this to the skill `PiiDetectionSkill`.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs [hunk 4]
```diff
@@ -1702,6 +1790,11 @@ public partial class SearchIndexerDataSourceConnection
         public static bool operator !=(Azure.Search.Documents.Indexes.Models.SearchIndexerDataSourceType left, Azure.Search.Documents.Indexes.Models.SearchIndexerDataSourceType right) { throw null; }
         public override string ToString() { throw null; }
     }
+    public partial class SearchIndexerDataUserAssignedIdentity : Azure.Search.Documents.Indexes.Models.SearchIndexerDataIdentity
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:1795]: I think these are independent.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:1755]: Agree. I could not think of a quick way to improve this part w/o a lot of churn. We should have a follow-up update to this.

File: sdk/search/Azure.Search.Documents/src/Options/SearchOptions.cs [hunk 5]
```diff
@@ -235,6 +237,61 @@ internal string QueryAnswerRaw
...
+                {
+                    QueryCaption = null;
+                    QueryCaptionHighlight = null;
+                }
+                else
+                {
+                    if (value.Contains(QueryCaptionRawSplitter))
+                    {
+                        var queryCaptionPart = value.Substring(0, value.IndexOf(QueryCaptionRawSplitter, StringComparison.OrdinalIgnoreCase));
+                        var highlightPart = value.Substring(value.IndexOf(QueryCaptionRawSplitter, StringComparison.OrdinalIgnoreCase) + QueryCaptionRawSplitter.Length);
```
Comment (tg-msft) [inline sdk/search/Azure.Search.Documents/src/Options/SearchOptions.cs:281]: Nit - 
```suggestion
                    int splitIndex = value.IndexOf(QueryCaptionRawSplitter, StringComparison.OrdinalIgnoreCase);
                    if (splitIndex >= 0)
                    {
                        var queryCaptionPart = value.Substring(0, splitIndex);
                        var highlightPart = value.Substring(splitIndex + QueryCaptionRawSplitter.Length);
```

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs (hunk 4 above)
[the remark this answers is not in the document]
Comment (tg-msft) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs:1795]: I think this is a user-assigned managed identity and we use things like `ManagedIdentityCredential` in `Azure.Identity`.  This is another one of those things we don't need to get right today but should revisit when Scott is back from vacation.

**Review by tg-msft — APPROVED** [review]
I left a few other comments that would be good to look into, but everything's good overall.

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs [hunk 6]
```diff
@@ -633,13 +636,41 @@ public partial class ElisionTokenFilter : Azure.Search.Documents.Indexes.Models.
...
+        public string DefaultLanguageCode { get { throw null; } set { } }
+        public double? MinimumPrecision { get { throw null; } set { } }
+        public string ModelVersion { get { throw null; } set { } }
+    }
     public partial class EntityRecognitionSkill : Azure.Search.Documents.Indexes.Models.SearchIndexerSkill
     {
         public EntityRecognitionSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs) { }
+        public EntityRecognitionSkill(System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.InputFieldMappingEntry> inputs, System.Collections.Generic.IEnumerable<Azure.Search.Documents.Indexes.Models.OutputFieldMappingEntry> outputs, Azure.Search.Documents.Indexes.Models.EntityRecognitionSkill.EntityRecognitionSkillVersion skillVersion) { }
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs]: That is the purpose of `EntityRecognitionSkill.SkillVersion` which is public. Or, do you mean something else?

File: sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs [hunk 7]
```diff
@@ -633,13 +636,41 @@ public partial class ElisionTokenFilter : Azure.Search.Documents.Indexes.Models.
...
         public bool? IncludeTypelessEntities { get { throw null; } set { } }
         public double? MinimumPrecision { get { throw null; } set { } }
+        public string ModelVersion { get { throw null; } set { } }
+        [System.Runtime.InteropServices.StructLayoutAttribute(System.Runtime.InteropServices.LayoutKind.Sequential)]
+        public readonly partial struct EntityRecognitionSkillVersion : System.IEquatable<Azure.Search.Documents.Indexes.Models.EntityRecognitionSkill.EntityRecognitionSkillVersion>
+        {
+            private readonly object _dummy;
+            private readonly int _dummyPrimitive;
+            public EntityRecognitionSkillVersion(string value) { throw null; }
+            public static Azure.Search.Documents.Indexes.Models.EntityRecognitionSkill.EntityRecognitionSkillVersion Latest { get { throw null; } }
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/api/Azure.Search.Documents.netstandard2.0.cs]: This is a client concept, in that, it does not flow to and from the service, but it allows users to create skills models using the versioned  properties in the unified model.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/EntityRecognitionSkill.cs [hunk 8]
```diff
@@ -1,14 +1,72 @@
...
+        /// <summary> The version of the model to use when calling the Text Analytics service.
+        /// It will default to the latest available when not specified.
+        /// We recommend you do not specify this value unless absolutely necessary. </summary>
+        public string ModelVersion
+        {
+            get => _modelVersion;
+
+            set
+            {
+                if (_skillVersion != EntityRecognitionSkillVersion.V3)
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/EntityRecognitionSkill.cs]: Without added support for `V4`+ it won't be very useful.
I have defined the operator that we can extend when further versions are added.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/EntityRecognitionSkillVersion.cs [hunk 9]
```diff
@@ -0,0 +1,62 @@
...
+
+            /// <summary> Creates a new instance of <see cref="EntityRecognitionSkillVersion"/>. </summary>
+            /// <exception cref="ArgumentNullException"> <paramref name="value"/> is null. </exception>
+            public EntityRecognitionSkillVersion(string value)
+            {
+                _value = value ?? throw new ArgumentNullException(nameof(value));
+            }
+
+            /// <summary> Version 1 of the <see cref="EntityRecognitionSkill"/>. </summary>
+            public static EntityRecognitionSkillVersion V1 { get; } = new EntityRecognitionSkillVersion(V1Value);
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/EntityRecognitionSkillVersion.cs]: I don't feel strongly about it. I implemented similar to what was reviewed for unified models (with Java code).
The good thing is we are not mixing it with OData type, the values can be changed down the line.

File: sdk/search/Azure.Search.Documents/src/Indexes/Models/OcrSkill.cs [hunk 10]
```diff
@@ -0,0 +1,21 @@
...
+
+namespace Azure.Search.Documents.Indexes.Models
+{
+    public partial class OcrSkill
+    {
+        /// <summary> A value indicating which language code to use. Default is <see cref="OcrSkillLanguage.En"/>. </summary>
+        public OcrSkillLanguage? DefaultLanguageCode { get; set; } = OcrSkillLanguage.En;
+
+        /// <summary> A value indicating to turn orientation detection on or not. Default is <c>false</c>. </summary>
+        public bool? ShouldDetectOrientation { get; set; } = false;
```
[the remark this answers is not in the document]
Comment (Mohit-Chakraborty) [reply sdk/search/Azure.Search.Documents/src/Indexes/Models/OcrSkill.cs:16]: I have seen many a times that the defaults that are called out are intended to be passed to the service.
I will take another look.

**Pull request merged**